Untangle docker

millerdev opened the pull request:
@snopoke

millerdev committed:
Incomplete, first pass at untangling docker setup

millerdev committed:
Fix travis/docker interactions

millerdev committed:
Add missing volumes

millerdev committed:
Optimize test setup

Eliminate python_env in favor of existing python environment in dimagi/commcarehq_base image. python_env was supposed to be an optimization for repeated runs on the same set of containers, but this is probably not the common case anyway, so not doing that for now.

millerdev committed:
Simplify docker script variables

millerdev committed:
Make docker ignore compose files (doesn't need them)

millerdev committed:
Do not setup ports when running tests in docker

millerdev committed:
Add better help for docker script

millerdev committed:
Make docker teardown command safer

millerdev committed:
Fix ES_DATA_VOLUME

millerdev committed:
Remove node_modules in volume

millerdev committed:
Clean up scripts/docker

millerdev committed:
Improve docker wait.sh and other minor cleanup

millerdev committed:
Get js tests running in docker

millerdev committed:
Add options for python tests on travis

millerdev committed:
Fix ./scripts/docker shell

millerdev committed:
Create kafka topics during docker test setup

millerdev committed:
Add ES_CLUSTER_NAME to elasticsearch docker service

millerdev committed:
Remove redundant ENV setting

millerdev committed:
Merge docker localsettings into travis localsettings

millerdev committed:
Add services --bootstrap option

millerdev committed:
Update Riak CS docker image

This fixes a bug that caused riak data to get broken when the container IP address changed.

millerdev committed:
Remove unneeded riak port

millerdev committed:
Fix riak data permissions issue on Mac

Some files in the riak data dir were owned as root. This was possibly due to a bug in Docker for Mac--it does not happen when using a docker data volume instead of a mounted host dir.

millerdev committed:
Add more examples to docker help

millerdev committed:
Run tests in docker as cchq rather than root

millerdev committed:
Remove old docker scripts and config files

millerdev committed:
Improve docker wait.sh

millerdev committed:
Move new docker scripts into docker dir

millerdev committed:
Remove travis matrix scripts

millerdev committed:
Move .travis/localsettings.py into docker dir

millerdev committed:
Update docker readme

millerdev committed:
Move commcarehq_base docker file

dimagimon commented:
There are 6 errors:

* docker/localsettings.py, line 132 - E251 unexpected spaces around keyword / parameter equals
* docker/localsettings.py, line 133 - E251 unexpected spaces around keyword / parameter equals
* docker/localsettings.py, line 136 - E251 unexpected spaces around keyword / parameter equals
* docker/localsettings.py, line 134 - E251 unexpected spaces around keyword / parameter equals
* corehq/apps/change_feed/management/commands/create_kafka_topics.py, line 1 - F401 'make_option' imported but unused
* docker/localsettings.py, line 135 - E251 unexpected spaces around keyword / parameter equals

snopoke commented on docker/mnt/run.sh:
```diff
@@ -0,0 +1,121 @@
+
+        # some kind of optimization?
+        # skip Running setup.py bdist_wheel for ... ?
+        rm -rf /root/.cache/pip
```
Only necessary if you're building an image.

snopoke commented on docker/mnt/run.sh:
```diff
@@ -0,0 +1,121 @@
+        pip install \
+            -r requirements/requirements.txt \
+            -r requirements/dev-requirements.txt \
+            coveralls
```
Doing this here vs during docker build means that you'll have to do it for every run. That's fine for travis but if you want to use it locally for anything then each run is going to be slower.

millerdev replied:
I've thought quite a bit about different options for this.
- One option is to put the virtualenv in a cached directory on the host and mount it into the container. That way it will not need to be rebuilt from scratch each time. I like this option best, but decided to defer implementing it until later because it's a little tricky to get the volume setup correctly.
- Currently since the dockerfile extends `dimagi/commcarehq_base:latest` there is already a not-too-outdated set of requirements installed when the container starts, so this doesn't take too long. This seemed like a reasonable stop-gap in the short term. Also, it might be a reasonable long-term solution if [`dimagi/commcarehq_base`](https://hub.docker.com/r/dimagi/commcarehq_base/) gets updated fairly frequently.

snopoke replied:
Yes, since the base image is built frequently the python env should be basically up to date (except for any changes in the current branch). I don't think caching the virtual env will make that much difference since you'll still need to run this same command on every build.

millerdev commented on corehq/apps/change_feed/management/commands/create_kafka_topics.py:
```diff
@@ -0,0 +1,21 @@
+        if client.has_metadata_for_topic(topic):
+            status = "already exists"
+        else:
+            client.ensure_topic_exists(topic, timeout=5)
```
This is a much faster way to create topics than the bash script it replaced. For this to work, Kafka [must be configured with `auto.create.topics.enable = true`](https://github.com/dpkp/kafka-python/issues/438#issuecomment-127815720) (that's [the default value](https://kafka.apache.org/08/configuration.html)). I didn't change anything in our kafka test setup to make this work.

millerdev commented on docker/.dockerignore:
```diff
@@ -0,0 +1,3 @@
+/*.yml
+/localsettings.py
+/Dockerfile_commcarehq_base
```
Is this dockerfile (path had been /docker/dockerfiles/Dockerfile_commcarehq_base) used to build [dimagi/commcarehq_base](https://hub.docker.com/r/dimagi/commcarehq_base/) on docker hub? It appears that image is build automatically on a schedule; is that correct? Do I need to change anything if I've moved the file within this repo?

snopoke replied:
Yes, it needs to by symlinked to the `Dockerfile` file in the repo root

millerdev commented on docker/compose/docker-compose-base.yml:
```diff
@@ -1,66 +0,0 @@
-    - "9200"
-
-web:
-  build: ../../
```
This was problematic because it sets the root of this dockerfile at the repo root, which means the entire repo (minus /.git, which is in /.dockerignore) is sent to the docker daemon when this image is built. This makes the build take a long time, and as far as I can tell we don't need to send all those files to the docker daemon.

snopoke commented on docker/mnt/run.sh:
```diff
@@ -0,0 +1,121 @@
+    fi
+    shift
+    setup $TEST
+    su cchq -c "../run_tests $TEST $(printf " %q" "$@")"
```
A but confused by the paths here. From what I can make out:
* This file is `/mnt/run.sh` 
* HQ source is in `/mnt/commcare-hq-ro`
* The `_run_tests` function get's written to `/mnt/run_tests` below
* `/mnt/lib` - not too sure what's in there or how HQ knows about it

Still trying to get head around the setup. Perhaps a quick overview would be helpful?

millerdev replied:
Yes, that's all correct.

The `_run_tests` thing is a hack that I don't like, but thought it was a good idea to run all non-setup code as a user other than root since the container is "[privileged](https://github.com/dimagi/commcare-hq/blob/943d45ddffe0d547d6b7defd355afd43105251a6/docker/hq-compose-bootstrap.yml#L6)" (which probably means it is root to the host OS).

`/mnt/lib` contains three directories: `node_modules`, `sharedfiles`, and `overlay`.

`overlay` is a read/write filesystem overlay for `/mnt/commcare-hq-ro`. An AUFS layered filesystem composed of `/mnt/lib/overlay` and `/mnt/commcare-hq-ro` is mounted at `/mnt/commcare-hq`. This allows the container to safely create and overwrite files in the repo without affecting anything outside the container.

millerdev replied:
Are there things that I could improve naming of or add comments to to make it clearer?

millerdev committed:
pep8

millerdev committed:
Add docker support for REUSE_DB

millerdev committed:
Use absolute paths for docker volumes on travis

millerdev committed:
Do not use read-only commcare-hq mount on travis

millerdev committed:
docker-compose 1.5.2 for variable substitution

millerdev committed:
Add missing Kafka topics to topics.ALL

millerdev committed:
Fix misc issues with docker setup

- Compile all pyc files for speed.
- Create empty staticfiles overlay with correct permissions so tests can write to it.
- Remove .log and .lock files so they can be recreated by test runner.
- Remove broken option used by docker teardown command.

dimagimon commented on corehq/apps/change_feed/topics.py:
```diff
@@ -11,10 +11,21 @@
 SMS = 'sms'
 LEDGER = 'ledger'
+COMMCARE_USER = 'commcare-user'
+GROUP = 'group
```
E901 SyntaxError: EOL while scanning string literal

snopoke commented on docker/hq-compose.yml:
```diff
@@ -0,0 +1,82 @@
+    POSTGRES_USER: commcarehq
+    POSTGRES_PASSWORD: commcarehq
+  labels:
+    commcare.name: "postgres"
```
These labels aren't necessary anymore

snopoke commented on docker/mnt/run.sh:
```diff
@@ -0,0 +1,129 @@
+    ln -s /mnt/lib/staticfiles lib/overlay/staticfiles
+    mount -t aufs -o br=lib/overlay:commcare-hq-ro none commcare-hq
+    chown cchq:cchq lib/overlay lib/staticfiles
+    rm commcare-hq/*.log commcare-hq/*.lock
```
If you make changes on the host will the propagate to docker without having to do a rebuild?

snopoke commented:
:+1: on :white_check_mark:

millerdev committed:
More misc issues

- Do not fail if log/lock files are missing.
- Compile custom and testapps (and hide output).
- Use pseudo-tty when running tests (fixes output issues with boot2docker).
- Write log file at path that can be written by cchq.

millerdev committed:
Add timeout to wait.sh in docker web container

millerdev committed:
Remove unused labels

millerdev committed:
Fix compress test: do not open template for write

millerdev committed:
Increase kafka topic creation timeout

millerdev committed:
Move coveralls out of docker (only used by travis)

millerdev committed:
Add 'web-user' kafka topic to topics.ALL

millerdev commented:
✅

millerdev commented on docker/README.md:
````diff
@@ -14,7 +14,7 @@ Initial setup
 
     ```
-      $ ./dockerhq.sh bootstrap
+      $ ./scripts/docker services --bootstrap
````
@dimagi/team-commcare-hq FYI the script to control docker in HQ is changing.

esoergel commented on docker/dockerfiles/Dockerfile_es:
```diff
@@ -1,4 +0,0 @@
-FROM elasticsearch:1.7.4
-
-# Install HEAD plugin
-RUN plugin -i mobz/elasticsearch-head
```
Does removing this file mean elasticsearch-head is no longer included in our docker containers?

millerdev replied:
Yes, sorry about that. Closing the loop for others who may find this: #12083 and #12128
